056-zk-kraft-migration.md

```diff
@@ -0,0 +1,247 @@
+The cluster has to be migrated to use `KafkaNodePool`(s) first for running the brokers (ZooKeeper nodes are never part of pools).
+Without them it’s not possible to deploy the new KRAft controllers (separated from the current Kafka brokers) as the first step of the migration.
+So the prerequisite is having the Kafka cluster still using ZooKeeper but running the Kafka brokers by using `KafkaNodePool`(s).
+The user can follow the procedure described [here](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str) in the official documentation, in order to meet this prerequisite.
+
+Also, the minimum Kafka version should be 3.5.0 with `inter.broker.protocol.version` set to “3.5”.
+The user could need to upgrade the Kafka version on its own cluster before running a migration.
+
+### Migration
+
+The migration process is semi-automatic which means having the user driving the process by interacting with the operator (mostly via annotations and custom resources creation/deletion) which is in charge to do the rest (i.e. nodes re-configuration, rolling, ...).
+
+> NOTE: this kind of approach is already in place for the Kafka version upgrade.
+
+Assume the following prerequisites, with a ZooKeeper based Kafka cluster defined by:
+
+* A `Kafka` custom resource.
+* One or more `KafkaNodePool`(s) for running the ZooKeeper based brokers.
+
+It would assume the operator is not using KRaft by default nor has the `+UseKRaft` feature gate enabled.
+
+First step for the user would be enabling the `+UseKRaft` feature gate unless it’s already enabled by default in a new operator release (depending on which Strimzi version this proposal will land).
+So it could be needed doing an upgrade of the current operator to the new release.
+In both cases, the operator is rolled, a new one starts and it’s able to handle the creation of new Apache Kafka clusters in KRaft mode but not migrating the old ZooKeeper based ones automatically.
+
+Currently, the operator prints the following warning:
+
+```shell
+io.strimzi.operator.common.model.InvalidResourceException: Tha Kafka cluster my-cluster is invalid: [At least one KafkaNodePool with the controller role and at least one replica is required when KRaft mode is enabled]
+```
+
+This is because with KRaft mode enabled in the operator, the reconciled `Kafka` resource doesn’t have any controller nodes (in any pool).
+
+The migration process can be also rolled back to ZooKeeper at any point before taking the KRaft controllers out of the migration mode.
+
+### Finite State Machine (FSM)
+
+The migration process can be described with a Finite State Machine (FSM).
+The operator has a new `MigrationManager` component in charge of tracking the migration finite state machine.
+Transitioning across the states depends on the current state, stored in the `Kafka.status.migrationState` field, and the "input" represented by the user interaction with the operator through the following annotations:
```

Comment:
This seems like a very permanent API change for a very temporary thing.

Reply:
This is a good point. We could use a `strimzi.io/kraft-migration-state` annotation bringing the current state instead.